File: suites/12_game_ai/comparisons/astar_pathfinding.py
def astar_pathfinding(grid_size: int) -> int:
    max_nodes = 10000
    checksum = 0
    
    for path_finds in range(100):
        start_x = (path_finds * 17) % grid_size
        start_y = (path_finds * 19) % grid_size
        goal_x = (grid_size - 1) - (path_finds * 23) % grid_size
        goal_y = (grid_size - 1) - (path_finds * 29) % grid_size
        
        nodes_expanded = 0
        curr_x, curr_y = start_x, start_y
        path_len = 0
        
        while (curr_x != goal_x or curr_y != goal_y) and nodes_expanded < max_nodes:
            nodes_expanded += 1
            best_next_x, best_next_y, best_f = curr_x, curr_y, 999999999
            
            for dy in range(-1, 2):
                for dx in range(-1, 2):
                    if dx != 0 or dy != 0:
                        nx, ny = curr_x + dx, curr_y + dy
                        if 0 <= nx < grid_size and 0 <= ny < grid_size:
                            is_obstacle = ((nx * 31 + ny * 37) % 100) < 20
                            if not is_obstacle:
                                g = path_len + 10 + (4 if dx != 0 and dy != 0 else 0)
                                h = (abs(nx - goal_x) + abs(ny - goal_y)) * 10
                                f = g + h
                                if f < best_f:
                                    best_f = f
                                    best_next_x, best_next_y = nx, ny
                                    
            if best_next_x == curr_x and best_next_y == curr_y:
                nodes_expanded = max_nodes # stuck
            else:
                curr_x, curr_y = best_next_x, best_next_y
                path_len += 10
                
        checksum = (checksum * 31 + path_len + nodes_expanded) % 1000000007
    return checksum

if __name__ == "__main__":
    print(astar_pathfinding(100))

File: suites/12_game_ai/comparisons/astar_pathfinding.go
package main

import "fmt"

func absVal(v int64) int64 {
	if v < 0 { return -v }
	return v
}

func astarPathfinding(gridSize int64) int64 {
	maxNodes := int64(10000)
	checksum := int64(0)

	for pathFinds := int64(0); pathFinds < 100; pathFinds++ {
		startX := (pathFinds * 17) % gridSize
		startY := (pathFinds * 19) % gridSize
		goalX := (gridSize - 1) - (pathFinds*23)%gridSize
		goalY := (gridSize - 1) - (pathFinds*29)%gridSize

		nodesExpanded := int64(0)
		currX, currY := startX, startY
		pathLen := int64(0)

		for (currX != goalX || currY != goalY) && nodesExpanded < maxNodes {
			nodesExpanded++
			bestNextX, bestNextY, bestF := currX, currY, int64(999999999)

			for dy := int64(-1); dy <= 1; dy++ {
				for dx := int64(-1); dx <= 1; dx++ {
					if dx != 0 || dy != 0 {
						nx, ny := currX+dx, currY+dy
						if nx >= 0 && nx < gridSize && ny >= 0 && ny < gridSize {
							isObstacle := ((nx*31+ny*37)%100 < 20)
							if !isObstacle {
								g := pathLen + 10
								if dx != 0 && dy != 0 { g += 4 }
								h := (absVal(nx-goalX) + absVal(ny-goalY)) * 10
								f := g + h
								if f < bestF {
									bestF = f
									bestNextX, bestNextY = nx, ny
								}
							}
						}
					}
				}
			}

			if bestNextX == currX && bestNextY == currY {
				nodesExpanded = maxNodes // stuck
			} else {
				currX, currY = bestNextX, bestNextY
				pathLen += 10
			}
		}
		checksum = (checksum*31 + pathLen + nodesExpanded) % 1000000007
	}
	return checksum
}

func main() { fmt.Println(astarPathfinding(100)) }
